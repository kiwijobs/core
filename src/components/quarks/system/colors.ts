import { get } from 'lodash';
import { system } from 'styled-system';
import { rgb } from '../../../theme/colors';
export const cc = (color: number, l: number, bg: number) => Math.floor(color * l + bg * (1 - l));
export const colorKeys = ['bg', 'backgroundColor', 'color', 'borderColor', 'fill', 'stroke'];

function hexify(color: string, lightness: number) {
  const toLight = lightness > 100;
  const hslAlpha = (toLight ? (lightness - 200) * -1 : lightness) / 100;
  const values = color
    ? color
        .replace(/rgba?\(/, '')
        .replace(/\)/, '')
        .replace(/[\s+]/g, '')
        .split(',')
        .map(c => cc(Number(c), hslAlpha, toLight ? 255 : 0))
        .join(', ')
    : '';

  return `rgb(${values})`;
}

export const transform = (value: any, scale: any) => {
  if (!value) return;

  const scaledValue = get(scale, value, '');
  if (scaledValue.length > 1) {
    return scaledValue;
  }

  if (value.indexOf('rgb') > -1 || value.indexOf('#') > -1) {
    return value;
  }

  const [colorOrValue, lightness] = value.split('.') || [];
  const color = scale[colorOrValue];

  if (!lightness) {
    return color || colorOrValue;
  }

  return hexify(color, lightness);
};

const config = {
  color: {
    property: 'color' as 'color',
    scale: 'colors' as 'colors',
    transform,
  },
  backgroundColor: {
    property: 'backgroundColor' as 'backgroundColor',
    scale: 'colors' as 'colors',
    transform,
  },
  opacity: true,
  bg: {},
};
config.bg = config.backgroundColor;

export const color = system(config);

export type TColor = keyof typeof rgb | number | string;

export interface ISystemColor {
  backgroundColor?: TColor;
  bg?: TColor;
  color?: TColor;
  borderColor?: TColor;
  fill?: TColor;
  stroke?: TColor;
}
