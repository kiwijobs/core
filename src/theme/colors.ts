import { IThemeValue } from '../theme';

export enum rgb {
  primary = 'rgb(8, 191, 130)',
  secondary = 'rgb(83, 109, 254)',
  tangerine = 'rgb(255, 151, 23)',
  pink = 'rgb(245, 0, 87)',
  white = 'rgb(255,255,255)',
  dark = 'rgb(40, 50, 56)',
}

export const colors: IThemeValue = [
  'inherit',
  'rgb(39,50,55)',
  'rgb(133,136,149)',
  'rgb(204,206,220)',
  'rgb(228,230,237)',
  'rgb(248,249,253)',
];
colors.dark = rgb.dark;
colors.white = rgb.white;
colors.primary = rgb.primary;
colors.secondary = rgb.secondary;
colors.tangerine = rgb.tangerine;
colors.pink = rgb.pink;

export const rgba = (alpha = 1) =>
  Object.entries(rgb).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]: value.replace(')', `, ${alpha})`).replace(/rgb/, 'rgba'),
    }),
    {} as typeof rgb
  );
