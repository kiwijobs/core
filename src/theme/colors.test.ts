import * as c from './colors';

describe('colors', () => {
  it('Has defined "rbg" colors', () => {
    expect(c.rgb).toMatchSnapshot();
  });

  it('Has defined colors for "styled-system"', () => {
    expect(c.colors).toMatchSnapshot();
  });

  it('Has working "rgba" function', () => {
    expect(typeof c.rgba).toBe('function');
    expect(c.rgba(0.4).dark).toEqual('rgba(0, 0, 0, 0.4)');
    expect(c.rgba().primary).toEqual('rgba(8, 191, 130, 1)');
  });
});

