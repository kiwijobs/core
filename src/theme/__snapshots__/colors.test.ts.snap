// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colors Has defined "rbg" colors 1`] = `
Object {
  "dark": "rgb(40, 50, 56)",
  "pink": "rgb(245, 0, 87)",
  "primary": "rgb(8, 191, 130)",
  "secondary": "rgb(83, 109, 254)",
  "tangerine": "rgb(255, 151, 23)",
  "white": "rgb(255,255,255)",
}
`;

exports[`colors Has defined colors for "styled-system" 1`] = `
Array [
  "inherit",
  "rgb(39,50,55)",
  "rgb(133,136,149)",
  "rgb(204,206,220)",
  "rgb(228,230,237)",
  "rgb(248,249,253)",
]
`;
